src/GraphQL/Value.hs

```diff
@@ -33,69 +49,167 @@ module GraphQL.Value
   -- determine maximum object depth.
   arbitrary = sized genValue
 
--- | Generate an arbitrary scalar value.
-genScalarValue :: Gen Value
-genScalarValue = oneof [ ValueInt <$> arbitrary
-                       , ValueFloat <$> arbitrary
-                       , ValueBoolean <$> arbitrary
-                       , ValueString <$> arbitrary
-                       , ValueEnum <$> arbitrary
-                       , pure ValueNull
-                       ]
-
 -- | Generate an arbitrary value, with objects at most @n@ levels deep.
-genValue :: Int -> Gen Value
+genValue :: Arbitrary scalar => Int -> Gen (Value' scalar)
 genValue n
-  | n <= 0 = genScalarValue
-  | otherwise = oneof [ genScalarValue
-                      , ValueObject <$> genObject (n - 1)
-                      , ValueList . List <$> listOf (genValue (n - 1))
+  | n <= 0 = arbitrary
+  | otherwise = oneof [ ValueScalar' <$> arbitrary
+                      , ValueObject' <$> genObject (n - 1)
+                      , ValueList' . List' <$> listOf (genValue (n - 1))
                       ]
 
+-- | A GraphQL value which contains no variables.
+type Value = Value' ConstScalar
+
+-- TODO: These next two definitions are quite internal. We should move this
+-- mode to Internal and then re-export the bits that end-users will use.
+
+-- | A GraphQL value which might contain some variables, some of which may not
+-- have definitions.
+type UnresolvedVariableValue = Value' UnresolvedVariableScalar
+
+pattern ValueInt :: Int32 -> Value
+pattern ValueInt x = ValueScalar' (ConstInt x)
+
+pattern ValueFloat :: Double -> Value
+pattern ValueFloat x = ValueScalar' (ConstFloat x)
+
+pattern ValueBoolean :: Bool -> Value
+pattern ValueBoolean x = ValueScalar' (ConstBoolean x)
+
+pattern ValueString :: String -> Value
+pattern ValueString x = ValueScalar' (ConstString x)
+
+pattern ValueEnum :: Name -> Value
+pattern ValueEnum x = ValueScalar' (ConstEnum x)
+
+pattern ValueList :: forall t. List' t -> Value' t
+pattern ValueList x = ValueList' x
+
+pattern ValueObject :: forall t. Object' t -> Value' t
+pattern ValueObject x = ValueObject' x
+
+pattern ValueNull :: Value
+pattern ValueNull = ValueScalar' ConstNull
+
+-- | If a value is an object, return just that. Otherwise @Nothing@.
+toObject :: Value' scalar -> Maybe (Object' scalar)
+toObject (ValueObject' o) = pure o
+toObject _ = empty
+
+-- * Scalars
+
+-- | A non-variable value which contains no other values.
+data ConstScalar
+  = ConstInt Int32
+  | ConstFloat Double
+  | ConstBoolean Bool
+  | ConstString String
+  | ConstEnum Name
+  | ConstNull
+  deriving (Eq, Ord, Show)
+
+instance ToJSON ConstScalar where
+  toJSON (ConstInt x) = toJSON x
+  toJSON (ConstFloat x) = toJSON x
+  toJSON (ConstBoolean x) = toJSON x
+  toJSON (ConstString x) = toJSON x
+  toJSON (ConstEnum x) = toJSON x
+  toJSON ConstNull = Aeson.Null
+
+-- | A value which contains no other values, and might be a variable that
+-- might lack a definition.
```

Comment:
Spec reference is here: https://facebook.github.io/graphql/#VariableDefinition. I've updated the comment. I hope it makes it more clear. 

I thought about "binding" or using "bound" in the comment, but I'm unsure.

A variable is this:

```graphql
$variable
```

A variable definition is this:

```graphql
query foo($variable: Int) {
  ...
}
```

And a value can be provided for a variable with something like:

```json
{
  "variable": 42
}
```

At which stage is the binding?